File: SexyBiscuit.Engine/Rendering/Camera3D.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using SexyBiscuit.Engine.Core;

namespace SexyBiscuit.Engine.Rendering;

/// <summary>
/// 3D perspective or orthographic camera component.
/// Reads position/orientation from the Actor's Transform3D.
/// Tag the actor "MainCamera3D" to make it the static Main camera; "MainCamera" and
/// then any live camera are the fallbacks, in that order.
/// </summary>
public sealed class Camera3D : Component
{
    // -------------------------------------------------------------------------
    // Static registry
    // -------------------------------------------------------------------------
    private static readonly List<Camera3D> _all = new();

    /// <summary>
    /// The camera the local player looks through, set by <see cref="Gameplay.PlayerController"/>
    /// when the pawn it possesses carries one. While it is alive, enabled and on an active actor
    /// it is what <see cref="Main"/> returns, whatever the scene's tags say: the possessed
    /// player's view must win over a preview camera left in the level.
    /// </summary>
    public static Camera3D? PlayerView { get; set; }

    /// <summary>Drops every registered camera. A test that leaks one poisons the next.</summary>
    internal static void ClearAll()
    {
        _all.Clear();
        PlayerView = null;
    }

    /// <summary>
    /// The player's view camera when there is one, otherwise the best active Camera3D:
    /// one tagged "MainCamera3D", failing that one tagged "MainCamera", failing that
    /// whichever was added first.
    /// </summary>
    /// <remarks>
    /// The fallback chain is not politeness, it is parity. This used to return null unless
    /// something carried the exact "MainCamera3D" tag while the browser fell back to
    /// "MainCamera" and then to any camera at all, so the bundled <c>3D Scene</c> template —
    /// whose camera is tagged "MainCamera" — rendered in the browser and rendered nothing
    /// natively, because <c>RenderSystem3D.Render</c> returns immediately on a null camera.
    /// A world-space <c>UiCanvas</c> resolves its camera through this property and would have
    /// vanished the same way. <c>CameraParityTests</c> pins the order.
    /// </remarks>
    public static Camera3D? Main
    {
        get
        {
            if (PlayerView is { Enabled: true } view && view.Actor != null && view.Actor.IsActive && _all.Contains(view))
                return view;

            return Tagged("MainCamera3D") ?? Tagged("MainCamera") ?? FirstLive();

            static Camera3D? Tagged(string tag)
            {
                foreach (var cam in _all)
                    if (Live(cam) && cam.Actor!.Tag == tag) return cam;
                return null;
            }

            static Camera3D? FirstLive()
            {
                foreach (var cam in _all)
                    if (Live(cam)) return cam;
                return null;
            }

            static bool Live(Camera3D cam) => cam.Enabled && cam.Actor != null && cam.Actor.IsActive;
        }
    }

    // -------------------------------------------------------------------------
    // Projection properties
    // -------------------------------------------------------------------------
    /// <summary>Vertical field of view in degrees. Default 60°.</summary>
    public float FieldOfView { get; set; } = 60f;

    public float NearClip { get; set; } = 0.1f;
    public float FarClip  { get; set; } = 1000f;

    public bool  IsOrthographic { get; set; } = false;

    /// <summary>Half-height of the orthographic view volume in world units.</summary>
    public float OrthoSize { get; set; } = 5f;

    // -------------------------------------------------------------------------
    // Cached Transform3D
    // -------------------------------------------------------------------------
    private Transform3D? _t3d;

    /// <summary>The transform this camera views from, created on the actor if absent.</summary>
    public Transform3D GetTransform3D()
    {
        if (_t3d != null) return _t3d;
        _t3d = Actor.GetComponent<Transform3D>();
        if (_t3d == null)
            _t3d = Actor.AddComponent<Transform3D>();
        return _t3d;
    }

    // -------------------------------------------------------------------------
    // Lifecycle
    // -------------------------------------------------------------------------
    public override void Awake()
    {
        _all.Add(this);
    }

    public override void OnDestroy()
    {
        _all.Remove(this);
        if (ReferenceEquals(PlayerView, this)) PlayerView = null;
    }

    // -------------------------------------------------------------------------
    // View / Projection
    // -------------------------------------------------------------------------
    /// <summary>Returns the view matrix derived from this camera's Transform3D.</summary>
    public Matrix GetViewMatrix()
    {
        var t = GetTransform3D();
        return Matrix.CreateLookAt(t.Position, t.Position + t.Forward, t.Up);
    }

    /// <summary>Returns the projection matrix for the given aspect ratio.</summary>
    public Matrix GetProjectionMatrix(float aspectRatio)
    {
        if (IsOrthographic)
        {
            float halfH = OrthoSize;
            float halfW = halfH * aspectRatio;
            return Matrix.CreateOrthographicOffCenter(-halfW, halfW, -halfH, halfH, NearClip, FarClip);
        }

        return Matrix.CreatePerspectiveFieldOfView(
            MathHelper.ToRadians(FieldOfView),
            aspectRatio,
            NearClip,
            FarClip);
    }

    // -------------------------------------------------------------------------
    // Screen-to-world ray
    // -------------------------------------------------------------------------
    /// <summary>
    /// Unprojcts a screen-space position into a world-space ray.
    /// Returns (origin, direction) — direction is normalised.
    /// </summary>
    public (Vector3 origin, Vector3 direction) ScreenToWorldRay(Vector2 screenPos, GraphicsDevice gd)
        => ScreenToWorldRay(screenPos, gd.Viewport.Width, gd.Viewport.Height);

    /// <summary>
    /// A ray from the camera through a point on screen, for picking. The direction is
    /// normalised.
    /// </summary>
    /// <remarks>
    /// Takes the viewport as two numbers rather than a <see cref="GraphicsDevice"/> so it can
    /// be called with no device at all — which is what lets a headless test and the browser
    /// read the same fixture. The device overload is sugar over this one.
    /// <para>
    /// The arithmetic is the browser's, deliberately, rather than <c>Viewport.Unproject</c>.
    /// The two agree to about a pixel for an ordinary camera and drift apart at wide fields of
    /// view, and two implementations of one answer is exactly the thing a shared fixture
    /// cannot pin. One implementation, mirrored.
    /// </para>
    /// </remarks>
    public (Vector3 origin, Vector3 direction) ScreenToWorldRay(Vector2 screenPos, float viewportWidth, float viewportHeight)
    {
        float width  = MathF.Max(1f, viewportWidth);
        float height = MathF.Max(1f, viewportHeight);

        // Screen pixels into normalised device coordinates, where Y is flipped because
        // screen Y grows downwards and clip space Y grows upwards.
        float ndcX = screenPos.X / width * 2f - 1f;
        float ndcY = 1f - screenPos.Y / height * 2f;

        Transform3D t = GetTransform3D();
        float aspect  = width / height;

        if (IsOrthographic)
        {
            float halfHeight = OrthoSize;
            float halfWidth  = halfHeight * aspect;
            Vector3 offset   = t.Right * (ndcX * halfWidth) + t.Up * (ndcY * halfHeight);
            return (t.Position + offset, Normalised(t.Forward));
        }

        float tanHalfFov = MathF.Tan(MathHelper.ToRadians(FieldOfView) * 0.5f);

        Vector3 direction = t.Forward
                          + t.Right * (ndcX * tanHalfFov * aspect)
                          + t.Up    * (ndcY * tanHalfFov);

        return (t.Position, Normalised(direction));
    }

    // -------------------------------------------------------------------------
    // World-to-screen
    // -------------------------------------------------------------------------

    /// <summary>
    /// Projects a world point to screen pixels, or null when it is behind the camera.
    /// </summary>
    /// <remarks>
    /// The null is the point of it. A point behind the camera has a negative <c>w</c>, and the
    /// perspective divide flips it to the opposite side of the screen rather than hiding it —
    /// so a caller that projects without checking draws a nameplate for the enemy standing
    /// behind them. <see cref="Rendering.TextRenderer3D"/> guards this by hand with a view-space
    /// Z test; this returns null instead, which is what the browser has always done.
    /// </remarks>
    public Vector2? WorldToScreen(Vector3 worldPoint, float viewportWidth, float viewportHeight)
    {
        float width  = MathF.Max(1f, viewportWidth);
        float height = MathF.Max(1f, viewportHeight);

        Matrix vp = GetViewMatrix() * GetProjectionMatrix(width / height);

        float w = worldPoint.X * vp.M14 + worldPoint.Y * vp.M24 + worldPoint.Z * vp.M34 + vp.M44;
        if (w <= 0f) return null;

        float clipX = (worldPoint.X * vp.M11 + worldPoint.Y * vp.M21 + worldPoint.Z * vp.M31 + vp.M41) / w;
        float clipY = (worldPoint.X * vp.M12 + worldPoint.Y * vp.M22 + worldPoint.Z * vp.M32 + vp.M42) / w;

        return new Vector2(
            (clipX * 0.5f + 0.5f) * width,
            (1f - (clipY * 0.5f + 0.5f)) * height);
    }

    /// <summary>Projects a world point to screen pixels using the device's viewport.</summary>
    public Vector2? WorldToScreen(Vector3 worldPoint, GraphicsDevice gd)
        => WorldToScreen(worldPoint, gd.Viewport.Width, gd.Viewport.Height);

    private static Vector3 Normalised(Vector3 v)
        => v.LengthSquared() > 0f ? Vector3.Normalize(v) : v;
}

File: html5/src/rendering/Camera3D.js
// -----------------------------------------------------------------------------
// Camera3D — the 3D view.
//
// `Camera3D.main` prefers the camera the player is actually looking through:
// a possessed pawn's camera wins over a scene camera tagged MainCamera3D, which
// is what makes Play show the game's view rather than the editor's.
// -----------------------------------------------------------------------------

import { Component } from '../core/Component.js';
import { registerComponent } from '../core/TypeRegistry.js';
import { PropertyType as P } from '../core/PropertyTypes.js';
import { Vector3, Matrix4, DEG2RAD } from '../math/index.js';
import { Transform3D } from '../core/Transform3D.js';

/** The 3D view transform. */
export class Camera3D extends Component {
    static schema = {
        fieldOfView:    { type: P.Number, default: 60, min: 1, max: 179 },
        nearClip:       { type: P.Number, default: 0.1, min: 0.001 },
        farClip:        { type: P.Number, default: 1000 },
        isOrthographic: { type: P.Bool, default: false },
        orthoSize:      { type: P.Number, default: 5, min: 0.01 },
    };

    /** Every live 3D camera. */
    static all = [];

    /** The camera a possessed pawn owns. Set by the player controller. */
    static playerView = null;

    /** Drops every registered camera. A test that leaks one poisons the next. */
    static clearAll() {
        Camera3D.all.length = 0;
        Camera3D.playerView = null;
    }

    /**
     * The view the renderer draws through: the possessed pawn's camera, else one tagged
     * 'MainCamera3D', else one tagged 'MainCamera', else whichever was added first.
     *
     * The liveness check is not decoration -- a switched-off camera used to win this
     * election here while the C# twin skipped it, so a scene that disabled its menu camera
     * rendered through it in the browser and through the next one natively. CameraParityTests
     * and cameraMain.test.js pin the order and the check together.
     */
    static get main() {
        if (Camera3D.playerView && live(Camera3D.playerView)) return Camera3D.playerView;

        return Camera3D.all.find((c) => live(c) && c.actor.tag === 'MainCamera3D')
            ?? Camera3D.all.find((c) => live(c) && c.actor.tag === 'MainCamera')
            ?? Camera3D.all.find(live)
            ?? null;
    }

    constructor() {
        super();
        /** Vertical field of view, in degrees. */
        this.fieldOfView = 60;
        this.nearClip = 0.1;
        this.farClip = 1000;
        this.isOrthographic = false;
        /** Half the vertical extent of the view when orthographic. */
        this.orthoSize = 5;
    }

    awake() {
        Camera3D.all.push(this);
        // A camera needs a 3D transform; a scene file may only carry the 2D one.
        if (!this.actor.getComponent(Transform3D)) this.actor.addComponent(Transform3D);
    }

    onDestroy() {
        const i = Camera3D.all.indexOf(this);
        if (i >= 0) Camera3D.all.splice(i, 1);
        if (Camera3D.playerView === this) Camera3D.playerView = null;
    }

    /** This camera's 3D transform, added on awake if the actor lacked one. */
    getTransform3D() {
        return this.actor.getComponent(Transform3D) ?? this.actor.addComponent(Transform3D);
    }

    /** World space into view space. */
    getViewMatrix() {
        const t = this.getTransform3D();
        const eye = t.position;
        const target = Vector3.add(eye, t.forward);
        return Matrix4.lookAt(eye, target, t.up);
    }

    /** View space into clip space. */
    getProjectionMatrix(aspectRatio) {
        if (this.isOrthographic) {
            const halfHeight = this.orthoSize;
            const halfWidth = halfHeight * aspectRatio;
            return Matrix4.orthographic(
                -halfWidth, halfWidth, -halfHeight, halfHeight, this.nearClip, this.farClip);
        }
        return Matrix4.perspective(
            this.fieldOfView * DEG2RAD, aspectRatio, this.nearClip, this.farClip);
    }

    /**
     * A ray from the camera through a point on screen, for picking.
     *
     * @param {{x: number, y: number}} screenPoint Canvas pixels.
     * @param {number} viewportWidth
     * @param {number} viewportHeight
     * @returns {{origin: Vector3, direction: Vector3}}
     */
    screenToWorldRay(screenPoint, viewportWidth, viewportHeight) {
        // Screen pixels into normalised device coordinates, where Y is flipped
        // because canvas Y grows downwards and clip space Y grows upwards.
        const ndcX = (screenPoint.x / viewportWidth) * 2 - 1;
        const ndcY = 1 - (screenPoint.y / viewportHeight) * 2;

        const t = this.getTransform3D();
        const origin = t.position;

        if (this.isOrthographic) {
            const halfHeight = this.orthoSize;
            const halfWidth = halfHeight * (viewportWidth / viewportHeight);
            const offset = Vector3.add(
                Vector3.scale(t.right, ndcX * halfWidth),
                Vector3.scale(t.up, ndcY * halfHeight));
            return { origin: Vector3.add(origin, offset), direction: t.forward.normalize() };
        }

        const tanHalfFov = Math.tan(this.fieldOfView * DEG2RAD / 2);
        const aspect = viewportWidth / viewportHeight;

        const direction = Vector3.add(
            Vector3.add(
                t.forward,
                Vector3.scale(t.right, ndcX * tanHalfFov * aspect)),
            Vector3.scale(t.up, ndcY * tanHalfFov)).normalize();

        return { origin, direction };
    }

    /** Projects a world point to canvas pixels, or null when it is behind the camera. */
    worldToScreen(worldPoint, viewportWidth, viewportHeight) {
        const viewProjection = Matrix4.multiply(
            this.getViewMatrix(), this.getProjectionMatrix(viewportWidth / viewportHeight));

        const p = Vector3.from(worldPoint);
        const m = viewProjection.m;
        const w = m[3] * p.x + m[7] * p.y + m[11] * p.z + m[15];
        if (w <= 0) return null;

        const clipX = (m[0] * p.x + m[4] * p.y + m[8] * p.z + m[12]) / w;
        const clipY = (m[1] * p.x + m[5] * p.y + m[9] * p.z + m[13]) / w;

        return {
            x: (clipX * 0.5 + 0.5) * viewportWidth,
            y: (1 - (clipY * 0.5 + 0.5)) * viewportHeight,
        };
    }
}
/** Whether a camera is eligible to be `main`: switched on, attached and alive. */
function live(camera) {
    return camera.enabled === true
        && camera.actor != null
        && !camera.actor.isDestroyed
        && camera.actor.isActive;
}

registerComponent(Camera3D, { category: 'Rendering', summary: 'The 3D view transform.' });
